sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/XdsProtocolTransformer.java

```diff
@@ -0,0 +1,171 @@
+
+    private static final int EXPECT_LENGTH = 4;
+
+    private XdsProtocolTransformer() {
+    }
+
+    /**
+     * get the mapping between service name of k8s and cluster of istio
+     *
+     * @param clusters clusters
+     * @return mapping
+     */
+    public static Map<String, List<String>> getService2ClusterMapping(List<Cluster> clusters) {
+        Map<String, List<String>> nameMapping = new HashMap<>();
+        for (Cluster cluster : clusters) {
+            if (cluster == null) {
+                continue;
+            }
+            Optional<String> serviceNameFromCluster2 = getServiceNameFromCluster(cluster.getName());
+            if (!serviceNameFromCluster2.isPresent()) {
+                continue;
+            }
+            String serviceName = serviceNameFromCluster2.get();
+            nameMapping.computeIfAbsent(serviceName, k -> new ArrayList<>()).add(cluster.getName());
+        }
+        return nameMapping;
+    }
+
+    /**
+     * get the instance of one service by xds protocol
+     *
+     * @param loadAssignments eds data
+     * @return instance of service
+     */
+    public static Set<ServiceInstance> getServiceInstances(
+            List<ClusterLoadAssignment> loadAssignments) {
+        return loadAssignments.stream()
+                .filter(Objects::nonNull)
+                .flatMap(loadAssignment -> getServiceInstancesFromLoadAssignment(loadAssignment).stream())
+                .collect(Collectors.toSet());
+    }
+
+    private static Set<ServiceInstance> getServiceInstancesFromLoadAssignment(ClusterLoadAssignment loadAssignment) {
+        String clusterName = loadAssignment.getClusterName();
+        Optional<String> serviceNameOptional = getServiceNameFromCluster(clusterName);
+        if (!serviceNameOptional.isPresent()) {
+            return new HashSet<>();
+        }
+        String serviceName = serviceNameOptional.get();
+        return processClusterLoadAssignment(loadAssignment, serviceName, clusterName);
+    }
+
+    private static Set<ServiceInstance> processClusterLoadAssignment(ClusterLoadAssignment loadAssignment,
+            String serviceName, String clusterName) {
+        List<LocalityLbEndpoints> localityLbEndpointList = loadAssignment.getEndpointsList();
+        if (localityLbEndpointList == null) {
+            return new HashSet<>();
+        }
+        return localityLbEndpointList.stream()
+                .filter(Objects::nonNull)
+                .flatMap(localityLbEndpoints -> processLocalityLbEndpoints(localityLbEndpoints, serviceName,
+                        clusterName).stream())
+                .collect(Collectors.toSet());
+    }
+
+    private static Set<ServiceInstance> processLocalityLbEndpoints(LocalityLbEndpoints localityLbEndpoints,
+            String serviceName, String clusterName) {
+        List<LbEndpoint> lbEndpointsList = localityLbEndpoints.getLbEndpointsList();
+        if (lbEndpointsList == null) {
```

Comment:
It is recommended that collectionUtils.isEmpty be used for judgment.

Reply:
done